src/regdesk/printer_manager.js

```diff
@@ -0,0 +1,171 @@
+  #minorDropdown;
+  #dropdowns;
+
+  #printerType = LP2844;
+
+  /**
+   * Initializes a new instance of the PrinterManager class.
+   *
+   * @param {Navigator} nav - Browser Navigator to hook up events to.
+   * @param {PrinterDropdown} adultDropdown - The dropdown object managing the adult printer.
+   * @param {PrinterDropdown} minorDropdown - The dropdown object managing the Minor printer.
+   */
+  constructor(nav, adultDropdown, minorDropdown) {
+    this.#adultDropdown = adultDropdown;
+    this.#adultDropdown.addEventListener(PrinterDropdown.ConnectEventName, async (e) => this.handleRequestConnect(e));
+    this.#adultDropdown.addEventListener(PrinterDropdown.DisconnectEventName, async (e) => this.handleRequestDisconnect(e));
+
+    this.#minorDropdown = minorDropdown;
+    this.#minorDropdown.addEventListener(PrinterDropdown.ConnectEventName, async (e) => this.handleRequestConnect(e));
+    this.#minorDropdown.addEventListener(PrinterDropdown.DisconnectEventName, async (e) => this.handleRequestDisconnect(e));
+
+    // Convenience array
+    this.#dropdowns = [this.#adultDropdown, this.#minorDropdown];
+
+    nav.usb.addEventListener('connect', async (e) => this.handleConnectPrinter(e));
+    nav.usb.addEventListener('disconnect', async (e) => this.handleDisconnectPrinter(e));
+  }
+
+  /**
+   * Assign a printer to a dropdown
+   * @param {LP2844} printer
+   * @param {PrinterDropdown} dropdown
+   */
+  assignPrinter(printer, dropdown) {
+    // Find the first dropdown in order, hope it's the right one!
+    // TODO: Make this smarter and pull from a cache or something for assignment
+    // instead of just first past the post.
+    dropdown = dropdown || this.#dropdowns.find(d => !d.printer);
+    if (dropdown) {
+      dropdown.setPrinter(printer);
+    } else {
+      console.warn("Connected to a printer that we couldn't assign anywhere!");
+    }
+  }
+
+  /**
+   * Handle a recquest to add a printer to a dropdown.
+   * @param {CustomEvent} o - PrinterDropdown connect event
+   * @param {PrinterDropdown} o.target - The dropdown target.
+   */
+  async handleRequestConnect({ target: dropdown }) {
+    if (dropdown.printer) {
+      // Spurious event dispatch?
+      return;
+    }
+
+    this.pairPrinter(dropdown);
+  }
+
+  /**
+   *
+   * @param {Object} o - Event Object.
+   * @param {USBDevice} o.device - The USB device to connect to as a printer.
+   * @returns
+   */
+  async handleConnectPrinter({ device }) {
+    // If we already have a tracked printer we shouldn't try to add it.
+    if (this.#dropdowns.some((d) => d.printer?.device === device)) {
+      return;
+    }
+
+    let printer = new this.#printerType(device);
+    try {
+      await printer.connect();
+    } catch (e) {
+      // TODO: Display something more useful.
+      console.error("Failed to connect to printer: " + e);
+    }
+
+    return printer;
+  }
+
+  /**
+   * Disconnect a printer from the page.
+   * @param {Object} o - Container object
```

Comment:
Too many params?

Reply:
Parameter destructuring! I came across the jsDoc syntax in a third party project I was working on for work last week, apparently this is the right way to document [anonymous objects for parameter destructuring](https://stackoverflow.com/a/42898969/1309423).

Reply:
![image](https://user-images.githubusercontent.com/1441553/146657903-32bdfb8c-4fa3-4ab9-8a35-55859ff99216.png)

This is how it ends up showing up in VSCode, the type hints forward through.

Reply:
Huh, that will take a while for me to get used to. Sorry for so many comments on it.

Unrelated: how are you embedding those pictures?

Reply:
Drag and drop into the comment markdown field and it'll upload 'em.

I also use ShareX on my Windows machine for screenshot markup, that's how I do the boxes and things.